_posts/2021-01-04-unit-tests-and-concurrency.md

```diff
@@ -0,0 +1,135 @@
+
+## The Refactor
+
+This worked reasonably well, however I decided to try and optimise it further. 
+
+My current code would: 
+
+- start with a progress bar
+- do API call #1 
+- do API call #2 N times concurrently 
+- combine the results 
+- show the full list in the UI 
+
+What if we display the data as it becomes available. 
+
+- start with a progress bar 
+- do API call #1 
+- show the data from API call #1 in the UI 
+- do API calls #2 N times concurrently without displaying progress 
+- update the UI N times as the API calls for #2 are completing
+
+I was already using `ListAdapter` which has built-in diffing support. I can send an updated list and it will do it's magic to update the changed elements only.
+
+## First attempt
+
+The original code was written as a single Rx chain. It emits a Loading state, then does the API calls, combines the data and emits a Content state (or Error state in case of failure).
+I modified it to: emit a Content state containing the list of elements from API call #1 right after the call is done. The data from call #2 had some default values for the time being.
+After each invocation of call #2, get the current state, update the element and emit a new content state
+I ran the app, it looked fine. I ran my tests and they failed as expected. I updated the tests to include the intermittent states (the changes included more view states being emitted). Ran the tests again, two worked, one failed. Ran again, all green.
+
+The updated test:
+
+```kotlin
+@Test
+fun `execute - first call succeeds, one failed call to #2 - one full item, one item with default value`() {
+    val items = listOf("Item 1" to "https://item1.jpg", "Item 2" to "invalid url")
+    val service = successService(items)
+
+    val viewModel = ViewModel(service)
+
+    val expected = listOf(
+        ViewEntity("Item 1", "https://item1.jpg".toHttpUrl()),
+        ViewEntity("Item 2", null)
+    )
+    viewModel.state
+        .test()
+        .also { viewModel.execute() }
+        .awaitCount(4)  // (0) Loading, (1) Content, (2) Content (3) Content
+        .assertValueAt(3, Content(expected))
+}
+```
+
+the only changes in the test are: the argument to `awaitCount` went from 2 to 4 (to account for the intermittent `Content` values) and the final result is now at index 3.
+
+My refactoring made my tests flaky. Depending of the scheduling and the order of the requests for API call #2 I would get the correct or wrong result. My code had a concurrency issue.
+
+My state lived in a serialized `BehaviorSubject`. I would update the state by reading it from the subject, create an immutable copy of the updated state, write back the new state to the `Subject`. However this is NOT thread safe. Since the state updates happened on the computation pool, multiple threads could do it at the same time. This resulted in inconsistent state, the results of some API calls would get overwritten by the next call.
```

Comment:
> My state lived in a serialized `BehaviorSubject`. (...) However this is NOT thread safe.

For my clarification here, as I might have missed something. You stated "lived", past, but isn't this still the case after you refactor you just added the calls in parallel and now you have this "content" sealed hierarchy? 🤔 

> Since the state updates happened on the computation pool, multiple threads could do it at the same time.

Following my previous sentence, wouldn't this still be true? You fixed the tests but "potentially" this could still happen on production code depending on the requests... right?